mypy/semanal.py

```diff
@@ -2359,7 +2361,15 @@ def visit_unary_expr(self, expr: UnaryExpr) -> None:
 
     def visit_index_expr(self, expr: IndexExpr) -> None:
         expr.base.accept(self)
-        if refers_to_class_or_function(expr.base):
+        if isinstance(expr.base, RefExpr) and expr.base.kind == TYPE_ALIAS:
+            # Special form -- subcribing a generic type alias.
+            # Perform the type substitution and create a new alias.
+            res = analyze_type_alias(expr,
+                                     self.lookup_qualified,
+                                     self.lookup_fully_qualified,
+                                     self.fail)
+            expr.analyzed = TypeAliasExpr(res, line=expr.line, runtime=True)
```

Comment:
@gvanrossum I reorganized things a bit. Now line and column are stored (whenever possible) in ``TypeAliasExpr.type``  not in ``TypeAliasExpr`` itself, this allows nicer error reporting, see below.